ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/command/toolbar/selector/PanelSelectorPresenter.java

```diff
@@ -0,0 +1,191 @@
+/*******************************************************************************
+ * Copyright (c) 2012-2017 Codenvy, S.A.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.ide.command.toolbar.selector;
+
+import com.google.gwt.user.client.ui.AcceptsOneWidget;
+import com.google.inject.Inject;
+import com.google.web.bindery.event.shared.EventBus;
+import org.eclipse.che.ide.api.mvp.Presenter;
+import org.eclipse.che.ide.api.parts.PartStack;
+import org.eclipse.che.ide.api.parts.PartStackStateChangedEvent;
+import org.eclipse.che.ide.api.parts.PartStackType;
+import org.eclipse.che.ide.api.parts.Perspective;
+import org.eclipse.che.ide.api.parts.PerspectiveManager;
+
+import static org.eclipse.che.ide.api.parts.PartStack.State.HIDDEN;
+
+/**
+ * Presenter to manage Panel selector widget and perspective layout.
+ */
+public class PanelSelectorPresenter implements Presenter, PanelSelectorView.ActionDelegate {
+
+    private PanelSelectorView view;
+
+    private PerspectiveManager perspectiveManager;
+
+    @Inject
+    public PanelSelectorPresenter(PanelSelectorView view,
+                                  PerspectiveManager perspectiveManager,
+                                  EventBus eventBus) {
+        this.view = view;
+        this.perspectiveManager = perspectiveManager;
+
+        view.setDelegate(this);
+
+        eventBus.addHandler(PartStackStateChangedEvent.TYPE, new PartStackStateChangedEvent.Handler() {
+            @Override
+            public void onPartStackStateChanged(PartStackStateChangedEvent event) {
+                updateButtonState();
+            }
+        });
+    }
+
+    @Override
+    public void go(AcceptsOneWidget container) {
+        container.setWidget(view);
+    }
+
+    @Override
+    public void onButtonClicked() {
+        view.showPopup();
+    }
+
+    @Override
+    public void onSelectorLeftClicked() {
+        showPanels(true, false, false);
```

Comment:
It's not readable. Try to understand this piece of code after a few month. You won't understand.